src/main/java/com/proximyst/ban/service/impl/ImplMessageService.java

```diff
@@ -148,125 +118,140 @@ public ImplMessageService(final @NonNull IUserService userService,
+        .thenApply(
+            opt -> opt.orElseThrow(() -> new IllegalArgumentException("Target of punishment cannot be unknown.")))
+        .thenCombine(
+            this.userService.getUser(punishment.getPunisher())
+                .thenApply(opt ->
+                    opt.orElseThrow(() -> new IllegalArgumentException("Punisher of punishment cannot be unknown."))),
+            (target, punisher) -> ArrayUtils.append(placeholders,
+                "targetName", target.getUsername(),
+                "targetUuid", target.getUuid().toString(),
+
+                "punisherName", punisher.getUsername(),
+                "punisherUuid", punisher.getUuid().toString(),
+
+                "punishmentId", punishment.getId().orElse(-1L).toString(),
+                "punishmentDate", SimpleDateFormat.getDateInstance().format(punishment.getDate()),
+                "reason", punishment.getReason()
+                    .map(MiniMessage.get()::escapeTokens)
+                    .orElse("No reason specified"),
+                "punishmentType", punishment.getPunishmentType().name(),
+                "punishmentVerb", punishment.getPunishmentType().getVerbPastTense().map(this.cfg),
+
+                "expiry", !punishment.currentlyApplies()
+                    ? this.cfg.formatting.isLifted
+                    : punishment.isPermanent()
+                        ? this.cfg.formatting.never
+                        : DurationFormatUtils
+                            .formatDurationHMS(punishment.getExpiration() - System.currentTimeMillis()),
+                "duration", punishment.isPermanent()
+                    ? this.cfg.formatting.permanently
+                    : this.cfg.formatting.durationFormat
+                        .replace("<duration>",
+                            DurationFormatUtils.formatDurationWords(
+                                punishment.getExpiration() - System.currentTimeMillis(),
+                                false,
+                                false
+                            )))
+        )
+        .thenCompose(p -> this.formatMessage(messageKey, p));
   }
 
-  private @NonNull Component parseWithTarget(
-      final @NonNull String msg,
-      final @NonNull BanUser target) {
-    return MiniMessage.get().parse(
-        msg,
+  private @NonNull CompletableFuture<@NonNull Component> formatMessageFuture(final @NonNull String message,
+      final @Positive int futureCount, final @NonNull Object @NonNull ... placeholders) {
+    final CompletableFuture<?>[] futures = new CompletableFuture<?>[futureCount];
 
-        "targetName", target.getUsername(),
-        "targetUuid", target.getUuid().toString()
-    );
+    // We know how many futures there are, but we also need to actually find them.
+    int futureIdx = 0;
+    for (final Object placeholder : placeholders) {
+      if (placeholder instanceof CompletableFuture) {
+        futures[futureIdx++] = (CompletableFuture<?>) placeholder;
```

Comment:
Unsafe cast, not all objects are CompletableFutures :(. You should at least add a comment explaining how this cast is safe. The best "fix" is to create a `<T> T unsafeCast(instance, Class<T> type)` utility method somewhere, to which you can also add the SuppressWarning annotation, this way you suppress the smallest possible amount of code

Reply:
Look at the line above. This is a future.

Reply:
My bad. This should be rewritten though, since you're mixing multiple unrelated types into the same array, which completely nullifies Java's compile-time type checking.

Reply:
That's the point. We don't use Kotlin, so we can abuse its infix operator `to`, making any kind of type-safety either hard or extremely ugly. The only thing I'd be actually willing to do here is to ensure the types of the placeholder keys to be strings, but that too isn't really a big deal (as it does `String#valueOf` on everything).

Reply:
As far as I can tell, the `String#valueOf` call happens in a different method. Within the `formatMessageFuture` method, there is no guarantee that all elements of the `placeholders` array are actually Strings. 

A comment about this cast would be nice, explaining how it is assumed that the `placeholders` array can only contain Strings and Futures.